lib/util/exec/hermit.ts

```diff
@@ -0,0 +1,90 @@
+import os from 'os';
+import fs from 'fs-extra';
+import upath from 'upath';
+import { GlobalConfig } from '../../config/global';
+import { logger } from '../../logger';
+import { rawExec } from './common';
+import type { RawExecOptions } from './types';
+
+export function isHermit(): boolean {
+  const { binarySource } = GlobalConfig.get();
+  return binarySource === 'hermit';
+}
+
+async function statFile(f: string): Promise<fs.Stats | undefined> {
+  let exists: fs.Stats | undefined = undefined;
+  try {
+    exists = await fs.stat(f);
+  } catch (e) {
+    // not doing anything when file not exists for errors
+  }
+
+  return exists;
+}
+
+async function statHermit(
+  defaultCwd: string,
+  parts: string[]
+): Promise<fs.Stats | undefined> {
+  const hermitForCwd = upath.join(...[defaultCwd, ...parts, 'bin', 'hermit']);
+  logger.trace({ hermitForCwd }, 'looking up hermit');
+  return await statFile(hermitForCwd);
+}
+
+export async function findHermitCwd(cwd: string): Promise<string> {
+  const defaultCwd = GlobalConfig.get('localDir') ?? '';
+  const parts = cwd.replace(defaultCwd, '').split(upath.sep);
+  let exists: fs.Stats | undefined = undefined;
+
+  // search the current relative path until reach the defaultCwd
+  while (parts.length > 0) {
+    exists = await statHermit(defaultCwd, parts);
+    // on file found. break out of the loop
+    if (exists !== undefined) {
+      break;
+    }
+    // otherwise, continue searching in parent directory
+    parts.pop();
+  }
+
+  // search in defaultCwd
+  if (exists === undefined) {
+    exists = await statHermit(defaultCwd, parts);
+  }
+
+  if (exists === undefined) {
+    throw new Error(`hermit not found for ${cwd}`);
+  }
+
+  // once the path is found, return ${[path}/bin
+  // so that hermit runs with ./hermit
+  return upath.join(...[defaultCwd, ...parts, 'bin']);
+}
+
+export async function getHermitEnvs(
+  rawOptions: RawExecOptions
+): Promise<Record<string, string>> {
+  const cwd = (rawOptions.cwd ?? '').toString();
+  const hermitCwd = await findHermitCwd(cwd);
+  logger.debug({ cwd, hermitCwd }, 'fetching hermit environment variables');
+  // with -r will output the raw unquoted environment variables to consume
+  const hermitEnvResp = await rawExec('./hermit env -r', {
+    ...rawOptions,
+    cwd: hermitCwd,
+  });
+  const hermitEnvVars = hermitEnvResp.stdout
+    .split(os.EOL)
+    .reduce((acc: Record<string, string>, line): Record<string, string> => {
+      const trimmedLine = line.trim();
+      if (trimmedLine === '') {
+        return acc;
+      }
+      const equalIndex = trimmedLine.indexOf('=');
+      const name = trimmedLine.substring(0, equalIndex);
```

Comment:
i'm not familiar with this but could this `equalIndex` return -1? then the substring would fail with index out of bound?
if yes please wrap in try/catch/ or another check as you see fit.

Reply:
you could also use regex to resolve this without index out of bound issues

Reply:
nah I don't think so. the output of `hermit env -r` will always have a `=` sign.